diff -ru /tmp/binutils-2.37/bfd/config.bfd binutils-2.37/bfd/config.bfd
--- /tmp/binutils-2.37/bfd/config.bfd	2021-07-08 13:37:19.000000000 +0200
+++ binutils-2.37/bfd/config.bfd	2021-09-08 11:29:46.646315266 +0200
@@ -1173,6 +1173,12 @@
     targ_selvecs="riscv_elf32_vec riscv_elf64_vec riscv_elf32_be_vec riscv_elf64_be_vec"
     want64=true
     ;;
+
+  riscv64*-*-norostb*)
+	targ_defvec=riscv_elf64_vec
+	targ_selvecs="riscv_elf32_vec riscv_elf64_vec riscv_elf32_be_vec riscv_elf64_be_vec"
+	want64=true
+    ;;
 #endif
 
   rl78-*-elf)
diff -ru /tmp/binutils-2.37/config.sub binutils-2.37/config.sub
--- /tmp/binutils-2.37/config.sub	2021-07-08 13:37:19.000000000 +0200
+++ binutils-2.37/config.sub	2021-09-08 11:31:42.498977080 +0200
@@ -1737,6 +1737,8 @@
 		;;
 	none)
 		;;
+	norostb*)
+		;;
 	*)
 		echo Invalid configuration \`"$1"\': OS \`"$os"\' not recognized 1>&2
 		exit 1
diff -ru /tmp/binutils-2.37/gas/configure.tgt binutils-2.37/gas/configure.tgt
--- /tmp/binutils-2.37/gas/configure.tgt	2021-07-08 13:37:19.000000000 +0200
+++ binutils-2.37/gas/configure.tgt	2021-09-08 11:30:06.402428048 +0200
@@ -361,6 +361,7 @@
   pru-*-*)				fmt=elf ;;
 
   riscv*-*-*)				fmt=elf ;;
+  riscv*-*-norostb*)				fmt=elf ;;
 
   rx-*-linux*)				fmt=elf em=linux ;;
 
diff -ru /tmp/binutils-2.37/ld/configure.tgt binutils-2.37/ld/configure.tgt
--- /tmp/binutils-2.37/ld/configure.tgt	2021-07-08 13:37:20.000000000 +0200
+++ binutils-2.37/ld/configure.tgt	2021-09-08 11:31:28.850899066 +0200
@@ -771,6 +771,10 @@
 			targ_extra_emuls="elf32lriscv elf64briscv elf32briscv"
 			targ_extra_libpath=$targ_extra_emuls
 			;;
+riscv64*-*-norostb*)		targ_emul=elf64lriscv_norostb
+			targ_extra_emuls="elf32lriscv_norostb elf64lriscv elf32lriscv elf64briscv elf32briscv"
+			targ_extra_libpath=$targ_extra_emuls
+			;;
 rs6000-*-aix[5-9]*)	targ_emul=aix5rs6
 			targ_extra_ofiles=
 			;;
@@ -1038,6 +1042,10 @@
   NATIVE_LIB_DIRS='/lib /usr/lib /usr/pkg/lib /usr/local/lib'
   ;;
 
+*-*-norostb*)
+  NATIVE_LIB_DIRS='/lib /local/lib'
+  ;;
+
 *-*-freebsd*)
   NATIVE_LIB_DIRS='/lib /usr/lib /usr/local/lib'
   ;;
diff -ru /tmp/binutils-2.37/ld/Makefile.am binutils-2.37/ld/Makefile.am
--- /tmp/binutils-2.37/ld/Makefile.am	2021-07-08 13:37:20.000000000 +0200
+++ binutils-2.37/ld/Makefile.am	2021-09-08 11:31:07.758778528 +0200
@@ -435,6 +435,7 @@
 	eelf64lppc.c \
 	eelf64lppc_fbsd.c \
 	eelf64lriscv.c \
+	eelf64lriscv_norostb.c \
 	eelf64lriscv_lp64f.c \
 	eelf64lriscv_lp64.c \
 	eelf64briscv.c \
@@ -923,6 +924,7 @@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/eelf64lriscv.Pc@am__quote@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/eelf64lriscv_lp64f.Pc@am__quote@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/eelf64lriscv_lp64.Pc@am__quote@
+@AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/eelf64lriscv_norostb.Pc@am__quote@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/eelf64briscv.Pc@am__quote@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/eelf64briscv_lp64f.Pc@am__quote@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/eelf64briscv_lp64.Pc@am__quote@
diff -ruN /tmp/binutils-2.37/ld/emulparams/elf64lriscv_norostb.sh binutils-2.37/ld/emulparams/elf64lriscv_norostb.sh
--- /tmp/binutils-2.37/ld/emulparams/elf64lriscv_norostb.sh	1970-01-01 01:00:00.000000000 +0100
+++ binutils-2.37/ld/emulparams/elf64lriscv_norostb.sh	2021-09-08 11:29:16.762144724 +0200
@@ -0,0 +1,2 @@
+source_sh ${srcdir}/emulparams/elf64lriscv-defs.sh
+OUTPUT_FORMAT="elf64-littleriscv-norostb"
